src/main/java/gregtech/common/covers/filter/BaseFilterContainer.java

```diff
@@ -0,0 +1,191 @@
+package gregtech.common.covers.filter;
+
+import gregtech.api.util.IDirtyNotifiable;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.math.MathHelper;
+import net.minecraftforge.items.ItemStackHandler;
+
+import com.cleanroommc.modularui.api.widget.IWidget;
+import com.cleanroommc.modularui.screen.ModularPanel;
+import com.cleanroommc.modularui.value.sync.GuiSyncManager;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+
+public abstract class BaseFilterContainer extends ItemStackHandler {
+
+    private int maxTransferSize = 1;
+    private int transferSize;
+    private @Nullable IFilter currentFilter;
+    private @Nullable Runnable onFilterInstanceChange;
+    private final IDirtyNotifiable dirtyNotifiable;
+
+    protected BaseFilterContainer(IDirtyNotifiable dirtyNotifiable) {
+        super();
+        this.dirtyNotifiable = dirtyNotifiable;
+    }
+
+    @Override
+    public int getSlotLimit(int slot) {
+        return 1;
+    }
+
+    public void onFilterInstanceChange() {
+        dirtyNotifiable.markAsDirty();
+        if (onFilterInstanceChange != null) {
+            onFilterInstanceChange.run();
+        }
+    }
+
+    public void setOnFilterInstanceChange(@Nullable Runnable onFilterInstanceChange) {
+        this.onFilterInstanceChange = onFilterInstanceChange;
+    }
+
+    public final @NotNull ItemStack getFilterStack() {
+        return this.getStackInSlot(0);
+    }
+
+    @Override
+    public void setStackInSlot(int slot, @NotNull ItemStack stack) {
+        if (ItemStack.areItemStacksEqual(stack, getFilterStack()))
+            return;
+
+        if (stack.isEmpty()) {
+            setFilter(null);
+        } else if (isItemValid(stack)) {
+            setFilter(FilterTypeRegistry.getFilterForStack(stack));
+        }
+
+        super.setStackInSlot(slot, stack);
+    }
+
+    @Override
+    public boolean isItemValid(int slot, @NotNull ItemStack stack) {
+        return FilterTypeRegistry.isFilter(stack);
+    }
+
+    public boolean isItemValid(ItemStack stack) {
+        return isItemValid(0, stack);
+    }
+
+    // todo update stack for insert and extract, though that shouldn't be called normally
```

Comment:
Is this TODO still applicable?

Reply:
i actually don't remember what that todo is for, so i'm removing it